parachain/pallets/ethereum-beacon-client/src/lib.rs

```diff
@@ -466,11 +475,58 @@ pub mod pallet {
 				update.signature_slot,
 			)?;
 
+			if !config::IS_MINIMAL {
+				log::info!(target: "ethereum-beacon-client","💫 spec is not minimal, checking ancestry proof");
+
+				let block_root_hash: H256 =
+					merkleization::hash_tree_root_block_roots(update.block_roots.clone())
+						.map_err(|_| Error::<T>::BlockRootsHashTreeRootFailed)?
+						.into();
+
+				ensure!(
+					Self::is_valid_merkle_branch(
+						block_root_hash,
+						update.block_roots_proof,
+						config::BLOCK_ROOTS_INDEX,
+						config::BLOCK_ROOTS_INDEX,
+						update.finalized_header.state_root
+					),
+					Error::<T>::InvalidAncestryMerkleProof
+				);
+
+				log::info!(target: "ethereum-beacon-client","💫 ancestry proof passed!");
+
+				Self::process_and_store_block_roots(
+					update.block_roots,
+					update.finalized_header.slot,
+					block_root,
+				);
+			} else {
+				log::info!(target: "ethereum-beacon-client","💫 spec is minimal, not checking ancestry proof");
+			}
+
 			Self::store_finalized_header(block_root, update.finalized_header);
 
 			Ok(())
 		}
 
+		fn process_and_store_block_roots(
+			block_roots: BoundedVec<H256, <T>::MaxSlotsPerHistoricalRoot>,
+			slot: u64,
+			finalized_header_block_root: H256,
+		) {
+			let max_slots = T::MaxSlotsPerHistoricalRoot::get() as u64;
+			let start = slot - max_slots;
+			let end = slot;
+			for n in start..end {
```

Comment:
Performing multiple storage writes in a loop may be problematic from a performance/weight perspective. How many loop iterations are you expecting?

Reply:
Max `8192` at worst case. 😬 Is there a better way to do a bulk insert? I'll check anyway. This is a the first basic implementation of ancestry proofs, where I send the entire `block_roots` array to the beacon client because there isn't a way to get a proof for just the single `block_root` you want to prove. I figured it would be best to implement a stop-gap for the vulnerability and then see if I can get the single block root proof generation to work using a Prysm work.

To summarize:
This PR: sends all the `block_roots`, does the proof and then stores the expected block root per slot in storage.
Final solution: only send the blocks proofs I'm interested in, with a multi-proof.

Reply:
OK cool! I'll use most of what I've done so far, I'll just work on the multiproof bit then. There are two alternatives here, will explain on Monday.

Reply:
is `MMR` also an option here to verify the ancestry proof? something like what we do for Beefy in relaychain?